File: android/src/main/java/com/aliyun/sls/dart/AliyunLogDartSdkPlugin.java
package com.aliyun.sls.dart;

import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import androidx.annotation.NonNull;
import com.aliyun.sls.android.producer.Log;
import com.aliyun.sls.android.producer.LogProducerClient;
import com.aliyun.sls.android.producer.LogProducerConfig;
import com.aliyun.sls.android.producer.LogProducerResult;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * AliyunLogFlutterSdkPlugin
 */
public class AliyunLogDartSdkPlugin implements FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    private LogProducerConfig logProducerConfig;
    private LogProducerClient logProducerClient;
    private Context context;
    private static final Handler sHandler = new Handler(Looper.getMainLooper());

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        context = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "aliyun_sls/flutter_sdk");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "initProducer":
                initProducer(call, result);
                return;
            case "setEndpoint":
            case "setProject":
            case "setLogstore":
            case "setAccessKey":
            case "setSource":
            case "setTopic":
            case "addTag":
            case "updateConfiguration":
                updateConfiguration(call, result);
                return;
            case "addLog":
                addLog(call, result);
                return;
            case "destroy":
                destroy(call, result);
                return;
            default:
                result.notImplemented();
        }
    }

    private void updateConfiguration(MethodCall call, Result result) {
        if (null == logProducerConfig) {
            result.success(error(LogProducerResult.LOG_PRODUCER_INVALID, "LogProducerConfig is null"));
            return;
        }

        updateLogProducerConfig(call);
        result.success(success());
    }

    private void initProducer(MethodCall call, Result result) {
        try {
            logProducerConfig = new LogProducerConfig();
            updateLogProducerConfig(call);
            String error = initPersistent(call);
            if (!TextUtils.isEmpty(error)) {
                result.success((error(LogProducerResult.LOG_PRODUCER_INVALID, error)));
                return;
            }

            logProducerClient = new LogProducerClient(logProducerConfig, (i, s, s1, i1, i2) -> {
                sHandler.post(() -> channel.invokeMethod("on_send_done", new HashMap<String, Object>() {
                        {
                            put("code", i);
                            put("errorMessage", s1);
                            put("logBytes", i2);
                            put("compressedBytes", i2);
                        }
                    }
                ));
            });
            result.success(success());
        } catch (Throwable e) {
            result.success(error(LogProducerResult.LOG_PRODUCER_INVALID, e.getMessage()));
        }
    }

    private void updateLogProducerConfig(MethodCall call) {
        if (null == logProducerConfig) {
            return;
        }

        final String endpoint = optArgument(call, "endpoint", null);
        final String project = optArgument(call, "project", null);
        final String logstore = optArgument(call, "logstore", null);
        final String accessKeyId = optArgument(call, "accessKeyId", null);
        final String accessKeySecret = optArgument(call, "accessKeySecret", null);
        final String securityToken = optArgument(call, "securityToken", null);

        final Boolean debuggable = optArgument(call, "debuggable", null);
        final Integer connectTimeout = optArgument(call, "connectTimeout", null);
        final Integer sendTimeout = optArgument(call, "sendTimeout", null);
        final Integer ntpTimeOffset = optArgument(call, "ntpTimeOffset", null);
        final Integer maxLogDelayTime = optArgument(call, "maxLogDelayTime", null);
        final Boolean dropDelayLog = optArgument(call, "dropDelayLog", null);
        final Boolean dropUnauthorizedLog = optArgument(call, "dropUnauthorizedLog", null);

        final String source = optArgument(call, "source", null);
        final String topic = optArgument(call, "topic", null);
        final Map<String, String> tags = optArgument(call, "tags", null);

        final Integer packetLogBytes = optArgument(call, "packetLogBytes", null);
        final Integer packetLogCount = optArgument(call, "packetLogCount", null);
        final Integer packetTimeout = optArgument(call, "packetTimeout", null);
        final Integer maxBufferLimit = optArgument(call, "maxBufferLimit", null);

        // endpoint & ak
        if (null != endpoint) {
            logProducerConfig.setEndpoint(endpoint);
        }
        if (null != project) {
            logProducerConfig.setProject(project);
        }
        if (null != logstore) {
            logProducerConfig.setLogstore(logstore);
        }
        if (null != accessKeyId && null != accessKeySecret) {
            if (null != securityToken) {
                logProducerConfig.resetSecurityToken(accessKeyId, accessKeySecret, securityToken);
            } else {
                logProducerConfig.setAccessKeyId(accessKeyId);
                logProducerConfig.setAccessKeySecret(accessKeySecret);
            }
        }
        // basic options
        if (null != connectTimeout) {
            logProducerConfig.setConnectTimeoutSec(connectTimeout);
        }
        if (null != sendTimeout) {
            logProducerConfig.setSendTimeoutSec(sendTimeout);
        }
        if (null != ntpTimeOffset) {
            logProducerConfig.setNtpTimeOffset(ntpTimeOffset);
        }
        if (null != maxLogDelayTime) {
            logProducerConfig.setMaxLogDelayTime(maxLogDelayTime);
        }
        if (null != dropDelayLog) {
            logProducerConfig.setDropDelayLog(dropDelayLog ? 1 : 0);
        }
        if (null != dropUnauthorizedLog) {
            logProducerConfig.setDropUnauthorizedLog(dropUnauthorizedLog ? 1 : 0);
        }

        // source & topic & tags
        if (null != source) {
            logProducerConfig.setSource(source);
        }
        if (null != topic) {
            logProducerConfig.setTopic(topic);
        }
        if (null != tags) {
            for (Entry<String, String> entry : tags.entrySet()) {
                logProducerConfig.addTag(entry.getKey(), entry.getValue());
            }
        }

        // packet
        if (null != packetLogBytes) {
            logProducerConfig.setPacketLogBytes(packetLogBytes);
        }
        if (null != packetLogCount) {
            logProducerConfig.setPacketLogCount(packetLogCount);
        }
        if (null != packetTimeout) {
            logProducerConfig.setPacketTimeout(packetTimeout);
        }
        if (null != maxBufferLimit) {
            logProducerConfig.setMaxBufferLimit(maxBufferLimit);
        }

        // debuggable
        if (null != debuggable && debuggable) {
            logProducerConfig.logProducerDebug();
        }
    }

    private String initPersistent(MethodCall call) {
        final Boolean persistent = optArgument(call, "persistent", null);
        final Boolean persistentForceFlush = optArgument(call, "persistentForceFlush", null);
        final String persistentFilePath = optArgument(call, "persistentFilePath", null);
        final Integer persistentMaxFileCount = optArgument(call, "persistentMaxFileCount", null);
        final Integer persistentMaxFileSize = optArgument(call, "persistentMaxFileSize", null);
        final Integer persistentMaxLogCount = optArgument(call, "persistentMaxLogCount", null);

        if (null != persistent && persistent) {
            logProducerConfig.setPersistent(1);
            logProducerConfig.setPersistentForceFlush(null != persistentForceFlush && persistentForceFlush ? 1 : 0);

            logProducerConfig.setPersistentMaxFileCount(null != persistentMaxFileCount ? persistentMaxFileCount : 10);
            logProducerConfig.setPersistentMaxLogCount(
                null != persistentMaxLogCount ? persistentMaxLogCount : 64 * 1024);
            logProducerConfig.setPersistentMaxFileSize(
                null != persistentMaxFileSize ? persistentMaxFileSize : 1024 * 1024);

            if (TextUtils.isEmpty(persistentFilePath)) {
                return "persistent file path must not be null.";
            }

            String path = setupPersistentFilePath(persistentFilePath);
            if (null != path) {
                logProducerConfig.setPersistentFilePath(path);
            }
        }
        return null;
    }

    private String setupPersistentFilePath(String filePath) {
        if (null == context) {
            return null;
        }

        File path = new File(context.getFilesDir(), filePath);
        if (!path.exists()) {
            //noinspection ResultOfMethodCallIgnored
            path.mkdirs();
        }
        return new File(path, "dat").getAbsolutePath();
    }

    private void destroy(MethodCall call, Result result) {
        if (null == logProducerClient) {
            result.success(error(LogProducerResult.LOG_PRODUCER_INVALID, "LogProducerClient is null"));
            return;
        }

        logProducerClient.destroyLogProducer();
        logProducerClient = null;

        result.success(success());
    }

    private void addLog(MethodCall call, Result result) {
        if (null == logProducerClient) {
            result.success(error(LogProducerResult.LOG_PRODUCER_INVALID, "LogProducerClient is null"));
            return;
        }

        Map<String, Object> data = optArgument(call, "log", null);
        if (null == data) {
            result.success(error(LogProducerResult.LOG_PRODUCER_INVALID, "log is null"));
            return;
        }

        Log log = new Log();
        for (Entry<String, Object> entry : data.entrySet()) {
            log.putContent(entry.getKey(), String.valueOf(entry.getValue()));
        }

        LogProducerResult res = logProducerClient.addLog(log);
        if (res == LogProducerResult.LOG_PRODUCER_OK) {
            result.success(success());
        } else {
            result.success(error(res));
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
        if (null != logProducerClient) {
            logProducerClient.destroyLogProducer();
        }
    }

    private static <T> T optArgument(MethodCall call, String key, T def) {
        try {
            //noinspection unchecked
            return (T)call.argument(key);
        } catch (Throwable t) {
            return def;
        }
    }

    private static Map<String, Object> success() {
        return success(null);
    }

    private static Map<String, Object> success(Map<String, Object> data) {
        return createResultData(LogProducerResult.LOG_PRODUCER_OK, data, null);
    }

    private static Map<String, Object> error(LogProducerResult code) {
        return error(code, null);
    }

    private static Map<String, Object> error(LogProducerResult code, String error) {
        return createResultData(code, null, error);
    }

    private static Map<String, Object> createResultData(LogProducerResult code, Map<String, Object> data,
        String error) {
        return new HashMap<String, Object>() {
            {
                put("code", code.ordinal());
                put("data", null != data ? data : new HashMap<>());
                put("error", TextUtils.isEmpty(error) ? "" : error);
            }
        };
    }
}

File: ios/Classes/AliyunLogDartSdkPlugin.m
#import "AliyunLogDartSdkPlugin.h"
#import "AliyunLogArgumentUtils.h"
#import "AliyunLogProducer/AliyunLogProducer.h"


@interface AliyunLogDartSdkPlugin ()
//@property LogProducerConfig *logProducerConfig;
//@property LogProducerClient *logProducerClient;
@property NSMutableDictionary<NSString*, NSDictionary<NSString*, NSObject*>*> *instanceHandlers;
- (void) initProducer: (FlutterMethodCall *)call result: (FlutterResult)result;
- (void) updateLogProducerConfig: (FlutterMethodCall *)call config: (LogProducerConfig *)config;
- (void) addLog: (FlutterMethodCall *)call result: (FlutterResult)result;
- (void) onLogSendDone: (NSString *)logstore resultCode: (LogProducerResult)resultCode errorMessage: (NSString *)errorMessage logBytes: (NSInteger)logBytes compressedBytes: (NSInteger)compressedBytes;
- (NSString *) randomToken;
@end

@implementation AliyunLogDartSdkPlugin

- (instancetype)init
{
    self = [super init];
    if (self) {
        _instanceHandlers = [NSMutableDictionary dictionary];
    }
    return self;
}

+ (instancetype) initWithChannel: (FlutterMethodChannel *)channel {
    AliyunLogDartSdkPlugin *plugin = [[AliyunLogDartSdkPlugin alloc] init];
    plugin.channel = channel;
    return plugin;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"aliyun_sls/flutter_sdk"
                                     binaryMessenger:[registrar messenger]
    ];


    AliyunLogDartSdkPlugin* instance = [AliyunLogDartSdkPlugin initWithChannel:channel];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"initProducer" isEqualToString:call.method]) {
        [self initProducer:call result:result];
    } else if ([@"addLog" isEqualToString:call.method]){
        [self addLog:call result:result];
    } else if ([@"setEndpoint" isEqualToString:call.method] ||
               [@"setProject" isEqualToString:call.method] ||
               [@"setLogstore" isEqualToString:call.method] ||
               [@"setAccessKey" isEqualToString:call.method] ||
               [@"setSource" isEqualToString:call.method] ||
               [@"setTopic" isEqualToString:call.method] ||
               [@"addTag" isEqualToString:call.method] ||
               [@"updateConfiguration" isEqualToString:call.method]){
        [self updateConfiguration:call result:result];
    } else if ([@"destroy" isEqualToString:call.method]) {
        [self destroy:call result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void) initProducer: (FlutterMethodCall *)call result: (FlutterResult) result {
    LogProducerConfig *_logProducerConfig = [[LogProducerConfig alloc] initWithEndpoint:@"" project:@"" logstore:@""];
    [self updateLogProducerConfig:call config:_logProducerConfig];
    NSString *error = [self setupPersistent:call];
    if (error.length > 0) {
        return result([self error:error code:LogProducerInvalid]);
    }

    LogProducerClient *_logProducerClient = [[LogProducerClient alloc] initWithLogProducerConfig:_logProducerConfig callback:on_send_done_function_in_flutter userparams:self];

    NSString *token = [self randomToken];
    [_instanceHandlers setObject:@{@"config": _logProducerConfig, @"client": _logProducerClient} forKey:token];

    result([self success: @{@"token": token}]);
}

- (LogProducerConfig *) getLogProducerConfigByToken: (FlutterMethodCall *)call {
    NSString *token = [AliyunLogArgumentUtils stringValue:call key:@"token"];
    if (token.length <= 0) {
        return nil;
    }

    return (LogProducerConfig *)[_instanceHandlers objectForKey:token][@"config"];
}

- (LogProducerClient *) getLogProducerClientByToken: (FlutterMethodCall *)call {
    NSString *token = [AliyunLogArgumentUtils stringValue:call key:@"token"];
    if (token.length <= 0) {
        return nil;
    }

    return (LogProducerClient *)[_instanceHandlers objectForKey:token][@"client"];
}

- (void) updateLogProducerConfig: (FlutterMethodCall *)call config: (LogProducerConfig *)config {
    LogProducerConfig *_logProducerConfig = config;
    if (nil == _logProducerConfig) {
        _logProducerConfig = [self getLogProducerConfigByToken:call];
    }
    
    if (nil == _logProducerConfig || nil == call.arguments) {
        return;
    }

    NSString *endpoint = [AliyunLogArgumentUtils stringValue:call key:@"endpoint"];
    NSString *project = [AliyunLogArgumentUtils stringValue:call key:@"project"];
    NSString *logstore = [AliyunLogArgumentUtils stringValue:call key:@"logstore"];
    NSString *accessKeyId = [AliyunLogArgumentUtils stringValue:call key:@"accessKeyId"];
    NSString *accessKeySecret = [AliyunLogArgumentUtils stringValue:call key:@"accessKeySecret"];
    NSString *securityToken = [AliyunLogArgumentUtils stringValue:call key:@"securityToken"];

    BOOL debuggable = [AliyunLogArgumentUtils boolValue:call key:@"debuggable"];
    int connectTimeout = [AliyunLogArgumentUtils intValue:call key:@"connectTimeout"];
    int sendTimeout = [AliyunLogArgumentUtils intValue:call key:@"sendTimeout"];
    int ntpTimeOffset = [AliyunLogArgumentUtils intValue:call key:@"ntpTimeOffset"];
    int maxLogDelayTime = [AliyunLogArgumentUtils intValue:call key:@"maxLogDelayTime"];
    BOOL dropDelayLog = [AliyunLogArgumentUtils boolValue:call key:@"dropDelayLog"];
    BOOL dropUnauthorizedLog = [AliyunLogArgumentUtils boolValue:call key:@"dropUnauthorizedLog"];

    NSString *source = [AliyunLogArgumentUtils stringValue:call key:@"source"];
    NSString *topic = [AliyunLogArgumentUtils stringValue:call key:@"topic"];
    NSDictionary *tags = [AliyunLogArgumentUtils dictValue:call key:@"tags"];

    int packetLogBytes = [AliyunLogArgumentUtils intValue:call key:@"packetLogBytes"];
    int packetLogCount = [AliyunLogArgumentUtils intValue:call key:@"packetLogCount"];
    int packetTimeout = [AliyunLogArgumentUtils intValue:call key:@"packetTimeout"];
    int maxBufferLimit = [AliyunLogArgumentUtils intValue:call key:@"maxBufferLimit"];

    if (endpoint.length > 0) {
        [_logProducerConfig setEndpoint:endpoint];
    }
    if (project.length > 0) {
        [_logProducerConfig setProject:project];
    }
    if (logstore.length > 0) {
        [_logProducerConfig setLogstore:logstore];
    }
    if (accessKeyId.length > 0 && accessKeySecret.length > 0) {
        if (securityToken.length > 0) {
            [_logProducerConfig ResetSecurityToken:accessKeyId accessKeySecret:accessKeySecret securityToken:securityToken];
        } else {
            [_logProducerConfig setAccessKeyId:accessKeyId];
            [_logProducerConfig setAccessKeySecret:accessKeySecret];
        }
    }

    // basic options
    if (connectTimeout > 0) {
        [_logProducerConfig SetConnectTimeoutSec:connectTimeout];
    }
    if (sendTimeout > 0) {
        [_logProducerConfig SetSendTimeoutSec:sendTimeout];
    }
    if (ntpTimeOffset > 0) {
        [_logProducerConfig SetNtpTimeOffset:ntpTimeOffset];
    }
    if (maxLogDelayTime > 0) {
        [_logProducerConfig SetMaxLogDelayTime:maxLogDelayTime];
    }
    [_logProducerConfig SetDropDelayLog:dropDelayLog ? 1 : 0];
    [_logProducerConfig SetDropUnauthorizedLog:dropUnauthorizedLog ? 1 : 0];

    // source & topic & tags
    if (source.length > 0) {
        [_logProducerConfig SetSource:source];
    }
    if (topic.length > 0) {
        [_logProducerConfig SetTopic:topic];
    }
    if (nil != tags) {
        for (NSString *key in tags) {
            if (key.length > 0) {
                [_logProducerConfig AddTag:key value:[tags objectForKey:key]];
            }
        }
    }

    // packet
    if (packetLogBytes > 0) {
        [_logProducerConfig SetPacketLogBytes:packetLogBytes];
    }
    if (packetLogCount > 0) {
        [_logProducerConfig SetPacketLogCount:packetLogCount];
    }
    if (packetTimeout > 0) {
        [_logProducerConfig SetPacketTimeout:packetTimeout];
    }
    if (maxBufferLimit > 0) {
        [_logProducerConfig SetMaxBufferLimit:maxBufferLimit];
    }

    // debuggable
    if (debuggable) {
        [LogProducerConfig Debug];
    }
}

- (NSString *) setupPersistent: (FlutterMethodCall *)call {
    BOOL persistent = [AliyunLogArgumentUtils boolValue:call key:@"persistent"];
    BOOL persistentForceFlush = [AliyunLogArgumentUtils boolValue:call key:@"persistentForceFlush"];
    NSString *persistentFilePath = [AliyunLogArgumentUtils stringValue:call key:@"persistentFilePath"];
    int persistentMaxFileCount = [AliyunLogArgumentUtils intValue:call key:@"persistentMaxFileCount"];
    int persistentMaxFileSize = [AliyunLogArgumentUtils intValue:call key:@"persistentMaxFileSize"];
    int persistentMaxLogCount = [AliyunLogArgumentUtils intValue:call key:@"persistentMaxLogCount"];
    
    // persistent
    if (persistent) {
        if (persistentFilePath.length <= 0) {
            return @"persistent file path must not be null.";
        }
        LogProducerConfig *_logProducerConfig = [self getLogProducerConfigByToken:call];
        
        [_logProducerConfig SetPersistent:1];
        [_logProducerConfig SetPersistentForceFlush:persistentForceFlush ? 1 : 0];
        [_logProducerConfig SetPersistentMaxFileCount:persistentMaxFileCount > 0 ? persistentMaxFileCount : 10];
        [_logProducerConfig SetPersistentMaxFileSize:persistentMaxFileSize > 0 ? persistentMaxFileSize : 1024 * 1024];
        [_logProducerConfig SetPersistentMaxLogCount:persistentMaxLogCount > 0 ? persistentMaxLogCount : 64 * 1024];
        
        NSString *path = [self setupPersistentFilePath:persistentFilePath];
        [_logProducerConfig SetPersistentFilePath:path];
    }
    return nil;
}

- (NSString *) setupPersistentFilePath: (NSString *)filePath {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *path = [[paths lastObject] stringByAppendingFormat:@"/%@", filePath];
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    BOOL existed = [fileManager fileExistsAtPath:path];
    if (NO == existed) {
        [fileManager createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:nil];
    }
    
    return [path stringByAppendingString:@"/dat"];
}

- (void) addLog: (FlutterMethodCall *)call result: (FlutterResult)result {
    LogProducerClient *_logProducerClient = [self getLogProducerClientByToken:call];
    if (nil == _logProducerClient || nil == call.arguments) {
        result([self error:@"arguments is null or client is null" code:LogProducerInvalid]);
        return;
    }

    NSDictionary *data = [AliyunLogArgumentUtils dictValue:call key:@"log"];
    if (nil == data) {
        result([self error:@"log data is null" code:LogProducerInvalid]);
        return;
    }

    Log *log = [Log log];
    [log putContents:data];
    LogProducerResult res = [_logProducerClient AddLog:log];
    if (LogProducerOK == res) {
        result([self success]);
    } else {
        result([self error:res]);
    }
}

- (void) updateConfiguration: (FlutterMethodCall *)call result: (FlutterResult)result {
    LogProducerConfig *_logProducerConfig = [self getLogProducerConfigByToken:call];
    if (nil == _logProducerConfig) {
        result([self error:@"LogProducerConfig is null" code:LogProducerInvalid]);
        return;
    }

    [self updateLogProducerConfig:call config:_logProducerConfig];
    result([self success]);
}

- (void) destroy: (FlutterMethodCall *)call result: (FlutterResult)result {
    LogProducerClient *_logProducerClient = [self getLogProducerClientByToken:call];
    if (nil == _logProducerClient) {
        result([self error:@"LogProducerConfig is null" code:LogProducerInvalid]);
        return;
    }

    [_logProducerClient DestroyLogProducer];
    _logProducerClient = nil;

    result([self success]);
}

- (NSDictionary *) success {
    return [self success: nil];
}

- (NSDictionary *) success: (NSDictionary *) data {
    return [self createResultData:LogProducerOK data:data error:nil];
}

- (NSDictionary *) error: (LogProducerResult) code {
    return [self error: @"" code:code];
}

- (NSDictionary *) error: (NSString *)error code: (LogProducerResult)code {
    return [self createResultData:code data:nil error:error];
}

- (NSDictionary *) createResultData: (LogProducerResult)code data: (NSDictionary *)data error: (NSString *)error {
    return @{
        @"code": @((int)code),
        @"data": (nil == data ? @{} : data),
        @"error": (error.length > 0 ? error : @"")
    };
}

- (void) onLogSendDone: (NSString *)logstore resultCode: (LogProducerResult)resultCode errorMessage: (NSString *)errorMessage logBytes: (NSInteger)logBytes compressedBytes: (NSInteger)compressedBytes {
    [_channel invokeMethod:@"on_send_done"
                 arguments:@{
        @"code": @((int)resultCode),
        @"errorMessage": [errorMessage copy],
        @"logBytes": @(logBytes),
        @"compressedBytes": @(compressedBytes)
    }
    ];
}

- (NSString *) randomToken {
    NSUInteger min = 100000;
    NSUInteger max = 999999;
    NSUInteger randomNumber = arc4random_uniform((uint32_t)(max - min + 1)) + min;
    return [NSString stringWithFormat:@"%lu", randomNumber];
}

void on_send_done_function_in_flutter(const char * config_name, log_producer_result result, size_t log_bytes, size_t compressed_bytes, const char * req_id, const char * error_message, const unsigned char * raw_buffer, void *user_param) {
    AliyunLogDartSdkPlugin *plugin = (__bridge AliyunLogDartSdkPlugin *)user_param;
    if (nil == plugin) {
        return;
    }

    [plugin onLogSendDone:[NSString stringWithUTF8String:config_name]
               resultCode:result
             errorMessage:[NSString stringWithUTF8String:error_message]
                 logBytes:log_bytes
          compressedBytes:compressed_bytes
    ];
}
@end
